sfe/sfe.go

```diff
@@ -0,0 +1,301 @@
+		// the account specified in the claim. At this point we should wait
+		// for the RA to process the request before returning to the client,
+		// just in case the request fails.
+
+		// Success, the account has been unpaused.
+		unpauseSuccessful(response, claims.Subject)
+	} else {
+		unpauseInvalidRequest(response)
+	}
+}
+
+// unpauseInvalidRequest is a helper that displays a page indicating there's no
+// action for a Subscriber to take due to an invalid or lack of JWT in the data
+// object.
+func unpauseInvalidRequest(response http.ResponseWriter) {
+	renderTemplate(response, "unpause-noParams.html", nil)
+}
+
+// unpauseSuccessful is a helper that displays a success message to the
+// Subscriber if their account has been unpaused.
+func unpauseSuccessful(response http.ResponseWriter, accountID string) {
+	type tmplData struct {
+		ShouldUnpause bool
+		AccountID     string
+	}
+	renderTemplate(response, "unpause-post.html", tmplData{true, accountID})
+}
+
+// unpauseFailure is a helper that displays a failure message to the Subscriber
+// indicating that their account has failed to unpause. This should only be used
+// when the JWT fails to validate and other failures should use
+// unpauseInvalidRequest.
+func unpauseFailure(response http.ResponseWriter) {
+	type tmplData struct {
+		ShouldUnpause bool
+	}
+	renderTemplate(response, "unpause-post.html", tmplData{false})
+}
+
+// unpausePrepareToUnpause is a helper that displays a page containing a filled
+// out form containing a JWT created by the WFE. Upon accessing that link, the
+// Subscriber would click the "unpause my account" button to submit the form
+// back to the SFE.
+func unpausePrepareToUnpause(response http.ResponseWriter, unpausePath string, incomingJWT string) {
+	type tmplData struct {
+		UnpausePath string
+		JWT         string
+	}
+	renderTemplate(response, "unpause-params.html", tmplData{unpausePath, incomingJWT})
+}
+
+// Unpause allows a requester to unpause their account via a form present on the
+// page.
+func (sfe *SelfServiceFrontEndImpl) Unpause(response http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodHead:
+		return
+	case http.MethodGet:
+		sfe.getHelper(response, unpauseJWT(request.URL.Query().Get("jwt")))
+	case http.MethodPost:
+		sfe.postHelper(response, unpauseJWT(request.URL.Query().Get("jwt")))
+	default:
+		response.Header().Set("Access-Control-Allow-Methods", "GET, HEAD, POST")
+		response.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+}
+
+type sfeJWTClaims struct {
+	jwt.Claims
+
+	// Version is a custom claim used to mass invalidate existing JWTs by
+	// changing the API version via unpausePath.
+	Version string `json:"apiVersion,omitempty"`
+}
+
+// validateUnpauseJWTforAccount validates the signature and contents of an
+// unpauseJWT and verify that the its claims match a set of expected claims.
+// Passing validations returns the claims or an error.
```

Comment:
> Passing validations returns the claims or an error.

There's an ambiguity in "passing" - does it mean like passing a parameter, or succeeding?

Also, returning the full `sfeJWTClaims` returns more info than we need the rest of the package to have. What if instead we parse the `Subject` in here and return a registration ID? Then the `sfeJWTClaims` type could be moved inside this function.